File: src/MyServiceBus/Topology/ReceiveEndpointTransportTopology.cs
namespace MyServiceBus.Topology;

public sealed class ReceiveEndpointTransportTopology
{
    public ReceiveEndpointTransportTopology(
        string name,
        bool durable,
        bool temporary,
        ushort prefetchCount,
        IReadOnlyList<MessageBinding> bindings,
        IReadOnlyDictionary<string, object?>? transportOptions = null,
        int concurrentMessageLimit = 1)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(name);
        ArgumentNullException.ThrowIfNull(bindings);

        if (durable && temporary)
            throw new ArgumentException("A receive endpoint cannot be both durable and temporary.");

        if (bindings.Count == 0)
            throw new ArgumentException("A receive endpoint must have at least one binding.", nameof(bindings));

        ArgumentOutOfRangeException.ThrowIfLessThan(concurrentMessageLimit, 1);

        Name = name;
        Durable = durable;
        Temporary = temporary;
        PrefetchCount = prefetchCount;
        ConcurrentMessageLimit = concurrentMessageLimit;
        Bindings = bindings
            .Select(binding => new MessageBinding
            {
                MessageType = binding.MessageType,
                EntityName = binding.EntityName
            })
            .ToArray();
        TransportOptions = transportOptions is null
            ? null
            : new System.Collections.ObjectModel.ReadOnlyDictionary<string, object?>(
                new Dictionary<string, object?>(transportOptions, StringComparer.Ordinal));
    }

    public string Name { get; }
    public bool Durable { get; }
    public bool Temporary { get; }
    public ushort PrefetchCount { get; }
    public int ConcurrentMessageLimit { get; }
    public IReadOnlyList<MessageBinding> Bindings { get; }
    public IReadOnlyDictionary<string, object?>? TransportOptions { get; }
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/topology/ReceiveEndpointTransportTopology.java
package com.myservicebus.topology;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public record ReceiveEndpointTransportTopology(
        String name,
        boolean durable,
        boolean temporary,
        int prefetchCount,
        List<MessageBinding> bindings,
        Map<String, Object> transportOptions) {

    public ReceiveEndpointTransportTopology {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Receive endpoint name cannot be blank");
        }
        if (durable && temporary) {
            throw new IllegalArgumentException("A receive endpoint cannot be both durable and temporary");
        }
        if (prefetchCount < 0) {
            throw new IllegalArgumentException("Receive endpoint prefetch count cannot be negative");
        }
        if (bindings == null || bindings.isEmpty()) {
            throw new IllegalArgumentException("A receive endpoint must have at least one binding");
        }

        bindings = bindings.stream().map(binding -> {
            MessageBinding copy = new MessageBinding();
            copy.setMessageType(binding.getMessageType());
            copy.setEntityName(binding.getEntityName());
            return copy;
        }).toList();
        transportOptions = transportOptions == null
                ? null
                : Collections.unmodifiableMap(new LinkedHashMap<>(transportOptions));
    }
}
